plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -235,6 +199,60 @@ func (p *Plugin) admitPod(nodeName string, a admission.Attributes) error {
 	}
 }
 
+func (p *Plugin) admitPodCreate(nodeName string, a admission.Attributes) error {
+	// require a pod object
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+	}
+
+	// only allow nodes to create mirror pods
+	if _, isMirrorPod := pod.Annotations[api.MirrorPodAnnotationKey]; !isMirrorPod {
+		return admission.NewForbidden(a, fmt.Errorf("pod does not have %q annotation, node %q can only create mirror pods", api.MirrorPodAnnotationKey, nodeName))
+	}
+
+	// only allow nodes to create a pod bound to itself
+	if pod.Spec.NodeName != nodeName {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with spec.nodeName set to itself", nodeName))
+	}
+	if len(pod.OwnerReferences) > 0 {
+		if len(pod.OwnerReferences) > 1 {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with a single owner reference set to itself", nodeName))
+		}
+		owner := pod.OwnerReferences[0]
+		if owner.APIVersion != v1.SchemeGroupVersion.String() ||
+			owner.Kind != "Node" ||
+			owner.Name != nodeName {
```

Comment:
no validation of node uid?

Reply:
also ensure BlockOwnerDeletion is nil or false

Reply:
Added BlockOwnerDeletion check.

For UID, that means we need to add a node informer. Do we care enough about the UID being correct that it's worth adding that logic here? What would the consequences be of an attacker using a different UID?

Reply:
yes, it's worth it to avoid gc uid cache injection issues. it should be ~5 lines to plumb:

```diff
diff --git a/plugin/pkg/admission/noderestriction/admission.go b/plugin/pkg/admission/noderestriction/admission.go
index c38b644ae9b..2e655541351 100644
--- a/plugin/pkg/admission/noderestriction/admission.go
+++ b/plugin/pkg/admission/noderestriction/admission.go
@@ -72,6 +72,7 @@ type Plugin struct {
        *admission.Handler
        nodeIdentifier nodeidentifier.NodeIdentifier
        podsGetter     corev1lister.PodLister
+       nodesGetter    corev1lister.NodeLister
        // allows overriding for testing
        features featuregate.FeatureGate
 }
@@ -84,6 +85,7 @@ var (
 // SetExternalKubeInformerFactory registers an informer factory into Plugin
 func (p *Plugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
        p.podsGetter = f.Core().V1().Pods().Lister()
+       p.nodesGetter = f.Core().V1().Nodes().Lister()
 }
 
 // ValidateInitialization validates the Plugin was initialized properly
@@ -94,6 +96,9 @@ func (p *Plugin) ValidateInitialization() error {
        if p.podsGetter == nil {
                return fmt.Errorf("%s requires a pod getter", PluginName)
        }
+       if p.nodesGetter == nil {
+               return fmt.Errorf("%s requires a node getter", PluginName)
+       }
        return nil
 }

...

+               // get the existing node
+               existingNode, err := p.nodesGetter.Get(nodeName)
+               if errors.IsNotFound(err) {
+                       return fmt.Errorf("error looking up node %s to verify uid: %v", nodeName, err)
+               }
+               if err != nil {
+                       return admission.NewForbidden(a, fmt.Errorf("error looking up node %s to verify uid: %v", nodeName, err))
+               }
+
...
```